packages/contracts-bedrock/contracts/universal/ProxyAdmin.sol

```diff
@@ -0,0 +1,266 @@
+            require(success);
+            return abi.decode(returndata, (address));
+        } else if (proxyType == ProxyType.Chugsplash) {
+            // bytes4(keccak256("getOwner()")) == 0x
+            (bool success, bytes memory returndata) = address(proxy).staticcall(hex"893d20e8");
+            require(success);
+            return abi.decode(returndata, (address));
+        } else if (proxyType == ProxyType.ResolvedDelegate) {
+            address owner = addressManager.owner();
+            return owner;
+        }
+    }
+
+    /**
+     * @dev Changes the admin of `proxy` to `newAdmin`.
+     *      This contract must be the current admin of `proxy`.
+     *
+     * @param proxy    The proxy that will have its admin updated
+     * @param newAdmin The address of the admin to update to
+     */
+    function changeProxyAdmin(Proxy proxy, address newAdmin) external onlyOwner {
+        ProxyType proxyType = getProxyType(address(proxy));
+
+        if (proxyType == ProxyType.OpenZeppelin) {
+            proxy.changeAdmin(newAdmin);
+        } else if (proxyType == ProxyType.Chugsplash) {
+            L1ChugSplashProxy(payable(proxy)).setOwner(newAdmin);
+        } else if (proxyType == ProxyType.ResolvedDelegate) {
+            Lib_AddressManager(addressManager).transferOwnership(newAdmin);
+        }
+    }
+
+    /**
+     * @dev Upgrades `proxy` to `implementation`.
+     *      This contract must be the admin of `proxy`.
+     *
+     * @param proxy          The address of the proxy
+     * @param implementation The address of the implementation
+     */
+    function upgrade(Proxy proxy, address implementation) public onlyOwner {
+        ProxyType proxyType = getProxyType(address(proxy));
+
+        if (proxyType == ProxyType.OpenZeppelin) {
+            proxy.upgradeTo(implementation);
+        } else if (proxyType == ProxyType.Chugsplash) {
+            bytes memory code = address(proxy).code;
+            L1ChugSplashProxy(payable(proxy)).setStorage(
+                0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc,
+                bytes32(uint256(uint160(implementation)))
+            );
+        } else if (proxyType == ProxyType.ResolvedDelegate) {
+            string memory name = getProxyName(address(proxy));
+            Lib_AddressManager(addressManager).setAddress(name, implementation);
+        }
+    }
+
+    /**
+     * @dev Upgrades `proxy` to `implementation` and calls a function on the new implementation.
+     *      This contract must be the admin of `proxy`.
+     *
+     * @param proxy           The proxy to call
+     * @param implementation  The implementation to upgrade the proxy to
+     * @param data            The calldata to pass to the implementation
+     */
+    function upgradeAndCall(
+        Proxy proxy,
+        address implementation,
+        bytes memory data
+    ) external payable onlyOwner {
+        upgrade(proxy, implementation);
```

Comment:
If OZ, do upgrade and call, otherwise do the low level call + pass value

Reply:
in ci, make sure there is no function selector clash with L1std bridge and chugsplash proxy